File: src/main/java/minic/ui/workbench/MiniCPlaybackController.java
package minic.ui;

import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.util.Duration;
import minic.settings.MiniCSettings;
import minic.uiapi.UiControlResultDto;

import java.util.Objects;

/**
 * JavaFX 播放控制器。
 */
public final class MiniCPlaybackController {
    private final MiniCWorkbenchViewModel viewModel;
    private final boolean timelineEnabled;
    private Timeline timeline;

    /**
     * 创建播放控制器。
     *
     * @param viewModel UI 状态模型
     */
    public MiniCPlaybackController(MiniCWorkbenchViewModel viewModel) {
        this(viewModel, true);
    }

    MiniCPlaybackController(MiniCWorkbenchViewModel viewModel, boolean timelineEnabled) {
        this.viewModel = Objects.requireNonNull(viewModel, "viewModel");
        this.timelineEnabled = timelineEnabled;
        MiniCSettings.setFrameIntervalChangeListener(() -> {
            if (running()) {
                restartTimeline();
            }
        });
        viewModel.currentStateProperty().addListener((obs, oldState, newState) -> {
            if (running() && newState != null && "PAUSED".equals(newState.playbackMode())) {
                stopTimeline();
            }
        });
    }

    /**
     * 开始普通播放。
     */
    public void play() {
        viewModel.play();
        restartTimeline();
    }

    /**
     * 开始两倍速播放。
     */
    public void playFast() {
        viewModel.playFast();
        restartTimeline();
    }

    /**
     * 暂停播放。
     */
    public void pause() {
        stopTimeline();
        viewModel.pause();
    }

    /**
     * 跳转到下一编译环节。
     *
     * @return 控制结果
     */
    public UiControlResultDto nextStage() {
        stopTimeline();
        return viewModel.nextStage();
    }

    /**
     * 手动执行一次 tick。
     *
     * @return tick 结果
     */
    public UiControlResultDto tickOnce() {
        var state = viewModel.currentStateProperty().get();
        if (state != null && "PAUSED".equals(state.playbackMode())) {
            stopTimeline();
            return null;
        }
        UiControlResultDto result = viewModel.tick();
        if (viewModel.currentStateProperty().get() != null && !viewModel.currentStateProperty().get().canNext()) {
            stopTimeline();
        }
        return result;
    }

    /**
     * 当前是否有定时器在运行。
     *
     * @return 是否运行中
     */
    public boolean running() {
        return timeline != null && timeline.getStatus() == Timeline.Status.RUNNING;
    }

    private void restartTimeline() {
        stopTimeline();
        if (!timelineEnabled) {
            return;
        }
        long interval = viewModel.currentStateProperty().get() == null
                ? 1000
                : viewModel.currentStateProperty().get().frameIntervalMillis();
        timeline = new Timeline(new KeyFrame(Duration.millis(interval), event -> tickOnce()));
        timeline.setCycleCount(Timeline.INDEFINITE);
        timeline.play();
    }

    private void stopTimeline() {
        if (timeline != null) {
            timeline.stop();
            timeline = null;
        }
    }
}

File: uiweb/src/workbench/MiniCPlaybackController.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type { UiControlResultDto } from "../translation/uiapi";
import type { MiniCWorkbenchViewModel } from "./MiniCWorkbenchViewModel";

export const miniCPlaybackControllerMirror = {
  "javaPath": "src/main/java/minic/uilocal/workbench/MiniCPlaybackController.java",
  "webPath": "uiweb/src/workbench/MiniCPlaybackController.ts",
  "packageName": "minic.uilocal",
  "exportName": "MiniCPlaybackController",
  "kind": "class",
  "imports": [
    "java.util.Objects",
    "javafx.animation.KeyFrame",
    "javafx.animation.Timeline",
    "javafx.util.Duration",
    "minic.settings.MiniCSettings",
    "minic.uiapi.UiControlResultDto"
  ],
  "fields": [
    {
      "name": "timeline",
      "signature": "private Timeline timeline"
    },
    {
      "name": "timelineEnabled",
      "signature": "private final boolean timelineEnabled"
    },
    {
      "name": "viewModel",
      "signature": "private final MiniCWorkbenchViewModel viewModel"
    }
  ],
  "methods": [
    {
      "name": "nextStage",
      "signature": "nextStage()"
    },
    {
      "name": "pause",
      "signature": "pause()"
    },
    {
      "name": "play",
      "signature": "play()"
    },
    {
      "name": "playFast",
      "signature": "playFast()"
    },
    {
      "name": "restartTimeline",
      "signature": "restartTimeline()"
    },
    {
      "name": "running",
      "signature": "running()"
    },
    {
      "name": "stopTimeline",
      "signature": "stopTimeline()"
    },
    {
      "name": "tickOnce",
      "signature": "tickOnce()"
    }
  ]
} as const satisfies JavaMirrorFile;

export class MiniCPlaybackController {
  static readonly mirror = miniCPlaybackControllerMirror;

  readonly mirror = miniCPlaybackControllerMirror;

  private timer: number | null = null;

  constructor(
    private readonly viewModel: MiniCWorkbenchViewModel,
    private readonly timelineEnabled = true,
  ) {
    this.viewModel.currentStateProperty().subscribe((state) => {
      if (this.running() && state !== null && state.playbackMode === "PAUSED") {
        this.stopTimeline();
      }
    });
  }

  play(): void {
    this.viewModel.runInBackground(this.startPlayback(() => this.viewModel.play()), "播放失败");
  }

  playFast(): void {
    this.viewModel.runInBackground(this.startPlayback(() => this.viewModel.playFast()), "2x 播放失败");
  }

  pause(): void {
    this.stopTimeline();
    this.viewModel.runInBackground(this.viewModel.pause(), "暂停失败");
  }

  nextStage(): Promise<UiControlResultDto> {
    this.stopTimeline();
    return this.viewModel.nextStage();
  }

  async tickOnce(): Promise<UiControlResultDto | null> {
    const state = this.viewModel.currentStateProperty().get();
    if (state !== null && state.playbackMode === "PAUSED") {
      this.stopTimeline();
      return null;
    }
    const result = await this.viewModel.tick();
    const nextState = this.viewModel.currentStateProperty().get();
    if (nextState !== null && !nextState.canNext) {
      this.stopTimeline();
    }
    return result;
  }

  running(): boolean {
    return this.timer !== null;
  }

  dispose(): void {
    this.stopTimeline();
  }

  summary(): string {
    return `MiniCPlaybackController: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }

  private restartTimeline(): void {
    this.stopTimeline();
    if (!this.timelineEnabled) {
      return;
    }
    const state = this.viewModel.currentStateProperty().get();
    if (state === null || state.playbackMode === "PAUSED" || !state.canNext) {
      return;
    }
    const interval = state.frameIntervalMillis;
    this.timer = window.setInterval(() => {
      this.viewModel.runInBackground(this.tickOnce(), "播放 tick 失败");
    }, interval);
  }

  private async startPlayback(start: () => Promise<UiControlResultDto>): Promise<void> {
    this.stopTimeline();
    await start();
    this.restartTimeline();
  }

  private stopTimeline(): void {
    if (this.timer !== null) {
      window.clearInterval(this.timer);
      this.timer = null;
    }
  }
}

export default MiniCPlaybackController;
